### ISSUE TITLE
Display of available firmware is not filtered when using withModelRestricted

### ISSUE THREAD

**Issue by cdjackson** [eclipse/smarthome#4647]
PaperUI currently shows all firmware for a thing type and is not filtering on model if withModelRestricted is enabled. It is using the thing type resource, which directly returns all the firmware from a provider for the relevant thing type.

Since the filtering is done in the FirmwareUpdateService, and not in the FirmwareProvider, there doesn't appear to be any way to perform this filtering at present

**Comment by hsudbrock** [comment]
I had assumed that consumers of firmwares were supposed to go via the `FirmwareUpdateService` (apparently I assumed wrongly, since the `ThingTypeResource` is going via the `FirmwareRegistry`).

I think it makes sense to move the filtering to the `FirmwareRegistry` (in the `FirmwareUpdateService`, the filtering would then no longer be necessary, since it retrieves firmwares from the `FirmwareRegistry` as well). Do you agree, @cdjackson ?

**Comment by cdjackson** [comment]
> I think it makes sense to move the filtering to the FirmwareRegistry

Yes - I think this is part of the solution - the interface will need to add a method ```getFirmwares(ThingUID thingUID)```.

>  the filtering would then no longer be necessary

I'm not sure I understand this? The filtering is still needed somewhere to perform the model specific filter (or are we talking about different things?)

In addition, there should also be a new REST endpoint at ```thing/thinguid/firmwares``` to get the filtered list applicable to the thing, and PaperUI should then use this endpoint

**Comment by hsudbrock** [comment]
> I'm not sure I understand this? The filtering is still needed somewhere to perform the model specific filter (or are we talking about different things?)

I meant moving the filtering from the `FirmwareUpdateService` to the `FirmwareRegistry`: The filtering could then be done in the new method `getFirmwares(ThingUID thingUID)` in the `FirmwareRegistry` (which would filter out firmwares not suitable for the thing). Then the `FirmwareUpdateService` would no longer need to do the filtering itself, but could use the new method from the `FirmwareRegistry`.

I think that I would also add a method `getLatestFirmware(ThingUID thingUID)` to the `FirmwareRegistry` (and both `getFirmwares` and `getLatestFirmware` with an additional `locale` parameter as well), to be consistent with the existing methods in the `FirmwareRegistry` that retrieve firmwares based on a thing type.

Furthermore, I agree, there should also be a corresponding REST endpoint to be used in the PaperUI.

**Comment by cdjackson** [comment]
Sounds good to me :)
